File: client/src/models/Investor.ts
import type Admin from "./Admin";
import type Transaction from "./Transaction";

type InvestorJSON = {
  id?: string;
  name?: string;
  email?: string;
  password?: string;
  status?: string;
  admin_id?:string
  balance?: number;
  transactions?: unknown;
};
class Investor {
  #id: string;
  status: string;
  name: string;
  email: string;
  balance: number;
  #admin_ref: Admin | null;
  constructor(investorJSON: InvestorJSON, admin?: Admin) {
    this.#id = investorJSON.id || "";
    this.status = investorJSON.status || "";
    this.name = investorJSON.name || "Unknow Investor";
    this.balance = investorJSON.balance || 0;
    this.email = investorJSON.email || "";
    this.#admin_ref = admin || null;
  }
  get id(): string {
    return this.#id;
  }
  get admin(): Admin | null {
    return this.#admin_ref;
  }

  get transactions(): Transaction[] {
    const t = this.#admin_ref?.get_transactions_by_investor(this.#id) || [];
    return t.sort((a, b) => b.date.getTime() - a.date.getTime());
  }
}

export { type InvestorJSON };
export default Investor;

File: server/models/investor.go
package models

import (
	"fmt"

	"github.com/google/uuid"
)

type Investor struct {
	ID           uuid.UUID     `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	Status       Status        `gorm:"type:varchar(20);not null;default:'active';index" json:"status"`
	Name         string        `gorm:"default:''" json:"name"`
	Balance      int           `gorm:"column:balance;not null;default:0" json:"balance"`
	AdminID      uuid.UUID     `gorm:"type:uuid;index" json:"admin_id"` // Foreign Key column
	Admin        *Admin        `gorm:"constraint:OnUpdate:CASCADE,OnDelete:SET NULL;foreignKey:AdminID" json:"admin"`
	Transactions []Transaction `gorm:"foreignKey:InvestorID" json:"transactions"` // ← add this
}

func (Investor) TableName() string {
	return "investors"
}

func (i Investor) String() string {
	str := fmt.Sprintf("Investor(id=%s, name='%s', status='%s', admin_id='%s', transactions=%d", i.ID, i.Name, i.Status, i.AdminID, len(i.Transactions))
	var id uuid.UUID
	if i.Admin.ID != id {
		str += fmt.Sprintf(", admin=Admin(%s)", i.Admin.Name)
	}
	str += ")"
	return str
}
